src/backend/distributed/sql/udfs/citus_finalize_upgrade_to_citus11/latest.sql

```diff
@@ -0,0 +1,184 @@
+-- citus_finalize_upgrade_to_citus11() is a helper UDF ensures
+-- the upgrade to Citus 11 is finished successfully. Upgrade to
+-- Citus 11 requires all active primary worker nodes to get the
+-- metadata. And, this function's job is to sync the metadata to
+-- the nodes that does not already have
+-- once the function finishes without any errors and returns true
+-- the cluster is ready for running distributed queries from
+-- the worker nodes. When debug is enabled, the function provides
+-- more information to the user.
```

Comment:
Reading the description here I wonder if we could, during the upgrade, do the following:

- mark all nodes to should have metadata
- mark all nodes as not have the metadata synced.

Ideally the maintenance daemon will sync the metadata in an idempotent way to all workers after an upgrade like this.

Has this been thought of as an option?
What is the reason this doesn't work?

Reply:
There are ~3.5 reasons I haven't gone into that route:

1) We simplified maintenance daemon to only sync the node metadata changes. We left that logic for `citus_update_node` and `citus_disable_node`. 

In general, given that we always sync metadata, it didn't make sense to keep the whole metadata sync logic in the bg worker.

2) When there are lots of metadata to sync (such as lots of objects and/or partitioned tables), this operation blocks majority of the user activity. See the relatively modest example above, where it took ~2 minutes. 
2.5) During the upgrade, we have to call `fix_all_partition_shard_index_names()` which requires to block even READS on the partitioned tables. We might have done this step separately, but still having all in the same tx looks safer.

3) Given that this operation might fail due to unexpected missing objects (or existing objects?), making this operation in the background might cause infinite trial of metadata sync. I  

Overall, I want to make this step explicitly executed (on Marlin part of upgrade step) and even if this fails, everything is rollbacked and your cluster keeps operating without metadata sync. Once the issue is fixed, you can re-run the script.